src/common_upgrades/change_macros_in_xml.py

```diff
@@ -51,6 +67,39 @@ def __init__(self, file_access, logger):
         self._file_access = file_access
         self._logger = logger
 
+    def add_macro(self, ioc_name, macro_to_add, pattern, description="No description", default_value=None):
+        """
+        Add a macro with a specified name and value to all IOCs whose name begins with ioc_name, unless a macro
+        with that name already exists
+
+        Args:
+            ioc_name: Name of the IOC to add the macro to (e.g. DFKPS would add macros to DFKPS_01 and DFKPS_02)
+            macro_to_add: Macro class with desired name and value
+            pattern: Regex pattern describing what values the macro accepts e.g. "^(0|1)$" for 0 or 1
+            description: Description of macro purpose
+            default_value: An optional default value for the macro
+        Returns:
+            None
+        """
+        for path, ioc_xml in self._file_access.get_config_files(IOC_FILE):
+            for ioc in self.ioc_tag_generator(path, ioc_xml, ioc_name):
+                macros = ioc.getElementsByTagName("macros")[0]
+                if find_macro_with_name(macros, macro_to_add.name):
+                    break
+                new_macro = ioc_xml.createElement("macro")
+                new_macro.setAttribute("name", macro_to_add.name)
+                new_macro.setAttribute("value", macro_to_add.value)
+                new_macro.setAttribute("pattern", pattern)
+                new_macro.setAttribute("description", description)
+                if default_value:
+                    new_macro.setAttribute("hasDefault", "yes")
+                    new_macro.setAttribute("defaultValue", default_value)
+                else:
+                    new_macro.setAttribute("hasDefault", "no")
```

Comment:
The macros in the iocs.xml file of the config only have the attributes `name` and `value`. The other attributes added here are only needed in the `config.xml` file in the IOC.